src/core/Akka/Actor/CoordinatedShutdown.cs

```diff
@@ -152,6 +152,76 @@ public static CoordinatedShutdown Get(ActorSystem sys)
         public const string PhaseBeforeActorSystemTerminate = "before-actor-system-terminate";
         public const string PhaseActorSystemTerminate = "actor-system-terminate";
 
+
+
+        /// <summary>
+        /// Reason for the shutdown, which can be used by tasks in case they need to do
+        /// different things depending on what caused the shutdown. There are some
+        /// predefined reasons, but external libraries applications may also define
+        /// other reasons.
+        /// </summary>
+        public class Reason
```

Comment:
I like this!

Reply:
Actually this should be an enum.

Reply:
Disagree. If it needs properties and data that can be overridden in the future, for whatever reason, you can't have that extensibility with an `Enum`

Reply:
I.E. the `CustomReason` example used in the spec earlier illustrates one such case for that.

Reply:
Then you can easily change it into tagged union or just change from enum to class. It's time to stop putting enum-like types into singletons:

1. They blow up the code base.
2. They are hardly maintainable. You need a lot of boilerplate to keep things like equality, hash codes and string formatting in place.
3. They are not safe - you never know if you have covered all switch cases and no tool on earth will help you with that.
4. They are slow and memory-consuming (this don't matter much in this particular case).

If you can model some cases as enums, you should do that. This is not JVM, they won't be optimized automatically.

Reply:
None of the above is true in this particular case - it's a class that gets used once on shutdown. I don't think these are even true generally.

Reply:
I actually started this as an enum, but then I got to the custom reason and I got stuck.
To support custom reasons with enum, we'd need to cast it to another enum extended with the custom reasons and I didn't like that. Honestly, the current approach I also don't like so much